### PR TITLE
Properly handle tagged primitives in control flow analysis

### LINKED ISSUE #43502: Regression: Tagged types are incorrectly refined
# Bug Report

### 🔎 Search Terms

refine regression 4.3 refine never

### 🕗 Version & Regression Information

Worked with 4.2, stopped working with 4.3 beta

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=4.3.0-dev.20210402#code/C4TwDgpgBAEghgZwBZQLxQcATgSwHYDmUAZFAN4D6FSiSAXFNgK4QC+A3AFCcBmTeAY2A4A9nigEIweMgAyEQsCQAKGsgYykASgZ4mAWwBGELOU5QLUAPRWoAQSjHMUCDx4iswKAKQQBAayh8b0RoESZTPkFhMSCEKCYECAATRxAoSBEwABtoAHdcYUIoACkAZXNLHB4oZVBMmrUUAEJUdAADTFxCdq1Ky0YkLBE8qDwIUYBRLGGsZXaAFSQceOSRCAQ8AHIvbJERQOycf2g4KCbern6oLCkI8SaAOlzFJC5WTiA

### 💻 Code

```ts
type Hash = string & {__hash: true};

function getHashLength(hash: Hash): number {
    // A best effort check in case our function is used by people writing JS
    if (typeof hash !== `string`)
      throw new Error(`This doesn't look like a hash`);

    return hash.length;
}
```

### 🙁 Actual behavior

The `hash` variable is incorrectly refined to `never`.

### 🙂 Expected behavior

The `hash` variable should remain `Hash`, since the `typeof` check is coherent with `Hash`.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #43502.

### CODE REVIEW

**Review by DanielRosenwasser — APPROVED** [review]

**Review by weswigham — COMMENTED** [review]
This looks OK, but what about something like `T & {__tag: void} & string`? Should the `T`'s constraints still be contributing facts? As written they always do, which is... Maybe bad? If `T` is constrained to tags, I think the original issue recurs.

**Comment by ahejlsberg** [comment]
> but what about something like T & {__tag: void} & string? Should the T's constraints still be contributing facts?

Yeah, good point. I'll make sure that's handled too.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -22013,7 +22013,10 @@ namespace ts {
                 return reduceLeft((<UnionType>type).types, (facts, t) => facts | getTypeFacts(t), TypeFacts.None);
             }
             if (flags & TypeFlags.Intersection) {
-                return reduceLeft((<UnionType>type).types, (facts, t) => facts & getTypeFacts(t), TypeFacts.All);
+                // When an intersection contains a primitive type we ignore object type constituents as they are
+                // presumably type tags. For example, in string & { __kind__: "name" } we ignore the object type.
+                const containsPrimitive = maybeTypeOfKind(type, TypeFlags.Primitive);
+                return reduceLeft((<UnionType>type).types, (facts, t) => containsPrimitive && getBaseConstraintOrType(t).flags & TypeFlags.Object ? facts : facts & getTypeFacts(t), TypeFacts.All);
```
Comment (weswigham) [inline src/compiler/checker.ts:22019]: If the constraint is a union, I don't think this is quite going to do the right thing.

**Review by weswigham — APPROVED** [review]
A test with union/intersection constraints would be nice, but I think this looks good.

**Pull request merged**